[SPARK-8401] [Build] Scala version switching build enhancements

mallman opened the pull request:
These commits address a few minor issues in the Scala cross-version support in the build:
1. Correct two missing `${scala.binary.version}` pom file substitutions.
2. Don't update `scala.binary.version` in parent POM. This property is set through profiles.
3. Update the source of the generated scaladocs in `docs/_plugins/copy_api_dirs.rb`.
4. Factor common code out of `dev/change-version-to-*.sh` and add some validation. We also test `sed` to see if it's GNU sed and try `gsed` as an alternative if not. This prevents the script from running with a non-GNU sed.

This is my original work and I license this work to the Spark project under the Apache License.

Commit:
Replace jackson-module-scala_2.10 with
jackson-module-scala_${scala.binary.version}

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
@mallman Have a look at https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark please

CC @ScrapCodes

ScrapCodes commented:
I am Ok with 1.

Is there a reason we should restrict ourselves to GNU SED ? I suppose our scripts were posix complaint.

mallman commented:
@srowen Will create a Jira ticket.

@ScrapCodes This is what I get with (presumably BSD) sed on OS X:

```
[msa@Michaels-MacBook-Pro spark-1.4]$ ./dev/change-version-to-2.11.sh 
[msa@Michaels-MacBook-Pro spark-1.4]$ gst
On branch scala-versions
Your branch and 'vamp/scala-versions' have diverged,
and have 7 and 4 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   assembly/pom.xml
	modified:   bagel/pom.xml
	modified:   core/pom.xml
	modified:   dev/change-scala-version.sh
	modified:   docs/_plugins/copy_api_dirs.rb
	modified:   examples/pom.xml
	modified:   external/flume-sink/pom.xml
	modified:   external/flume/pom.xml
	modified:   external/kafka-assembly/pom.xml
	modified:   external/kafka/pom.xml
	modified:   external/mqtt/pom.xml
	modified:   external/twitter/pom.xml
	modified:   external/zeromq/pom.xml
	modified:   extras/java8-tests/pom.xml
	modified:   extras/kinesis-asl/pom.xml
	modified:   extras/spark-ganglia-lgpl/pom.xml
	modified:   graphx/pom.xml
	modified:   launcher/pom.xml
	modified:   mllib/pom.xml
	modified:   network/common/pom.xml
	modified:   network/shuffle/pom.xml
	modified:   network/yarn/pom.xml
	modified:   pom.xml
	modified:   repl/pom.xml
	modified:   sql/catalyst/pom.xml
	modified:   sql/core/pom.xml
	modified:   sql/hive-thriftserver/pom.xml
	modified:   sql/hive/pom.xml
	modified:   streaming/pom.xml
	modified:   tools/pom.xml
	modified:   unsafe/pom.xml
	modified:   yarn/pom.xml

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	assembly/pom.xml-e
	bagel/pom.xml-e
	core/pom.xml-e
	dev/audit-release/blank_maven_build/pom.xml-e
	dev/audit-release/maven_app_core/pom.xml-e
	docs/_plugins/copy_api_dirs.rb-e
	examples/pom.xml-e
	external/flume-sink/pom.xml-e
	external/flume/pom.xml-e
	external/kafka-assembly/pom.xml-e
	external/kafka/pom.xml-e
	external/mqtt/pom.xml-e
	external/twitter/pom.xml-e
	external/zeromq/pom.xml-e
	extras/java8-tests/pom.xml-e
	extras/kinesis-asl/pom.xml-e
	extras/spark-ganglia-lgpl/pom.xml-e
	graphx/pom.xml-e
	launcher/pom.xml-e
	mllib/pom.xml-e
	network/common/pom.xml-e
	network/shuffle/pom.xml-e
	network/yarn/pom.xml-e
	pom.xml-e
	repl/pom.xml-e
	sql/catalyst/pom.xml-e
	sql/core/pom.xml-e
	sql/hive-thriftserver/pom.xml-e
	sql/hive/pom.xml-e
	streaming/pom.xml-e
	tools/pom.xml-e
	unsafe/pom.xml-e
	yarn/pom.xml-e

no changes added to commit (use "git add" and/or "git commit -a")
```

mallman commented:
@srowen Should I create one Jira ticket for this or multiple?

srowen commented:
One JIRA is fine. Yes I see the same problem with OS X `sed`, so falling back to `gsed` is probably good.

mallman commented:
@srowen I created the Jira ticket which shows the problem with the current version changing scripts.

srowen commented on dev/change-scala-version.sh:
```diff
@@ -0,0 +1,63 @@
+}
+
+# Find GNU sed. On OS X with MacPorts you can install gsed with "sudo port install gsed"
+if test_sed sed; then
```
Can you use the `command -v` idiom seen elsewhere in our scripts to check for a binary?

mallman replied:
I can, but it seems that idiom is used to simply check for the existence of a binary. We want to test that `sed --version` outputs `GNU sed`. `command -v` would be inadequate for that purpose.

However, if you still want me to then I guess

    test_sed() {
      [ $(command -v $1) -a ! -z "$($1 --version 2>&1 | head -n 1 | grep 'GNU sed')" ]
    }

is the way to go? Or perhaps `test_gnu_sed` would be a better name for that function?

srowen replied:
Nah I see your point now about looking for GNU-ness. Sounds good.

srowen commented on dev/change-scala-version.sh:
```diff
@@ -0,0 +1,63 @@
+  | xargs -I {} $SED -i -e 's/\(artifactId.*\)_'$FROM_VERSION'/\1_'$TO_VERSION'/g' {}
+
+# Update source of scaladocs
+$SED -i -e 's/scala\-'$FROM_VERSION'/scala\-'$TO_VERSION'/' $BASEDIR/docs/_plugins/copy_api_dirs.rb
```
This line seems to have changed the logic though. It's supposed to update `scala.binary.version` right?

mallman replied:
It used to, yes. However, that property is now set through activating the appropriate maven profile—`scala-2.10` or `scala-2.11`. Changing it through a textual find/replace will override that logic. For example, the `scala-2.10` profile will compile with scala 2.11 or vice-versa.

srowen replied:
I don't think that will work, and I 80% remember why -- the published POM doesn't have any notion of activated profiles, and so the artifacts will default to expressing a dependency on 2.10. That's why it had to be hard-coded and changed on update to 2.11.

mallman replied:
That makes sense. I also better remember why I removed that line—it has to do with some incorrect behavior in changing `scala.binary.version` which I mistakenly attributed to that portion of the script. Absent any apparent problem, I'm going to restore that part of that script.

Commit:
Update source of scaladocs when changing Scala version

Commit:
Factor change-scala-version.sh out of change-version-to-*.sh, adding
command line argument validation and testing for GNU sed

srowen commented:
Looking good -- what's the need for the scaladoc update? wouldn't the build generate the scaladoc with the right location and name once the build is updated? Maybe there's a subtle reason there, but want to make sure this is something that must be done here and not elsewhere.

mallman commented:
Indeed the build does generate the scaladoc in the right location, but the `docs/_plugin/copy_api_dirs.rb` is currently hardcoded to always look for the api docs in `target/scala-2.10`:

https://github.com/apache/spark/blob/7af3818c6b2bf35bfa531ab7cc3a4a714385015e/docs/_plugins/copy_api_dirs.rb#L37

andrewor14 commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35166 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35166/consoleFull) for   PR 6832 at commit [`ad9b40a`](https://github.com/apache/spark/commit/ad9b40acf7daa28ba9b7faff279cfbad4b00c096).

SparkQA commented:
[Test build #35166 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35166/console) for   PR 6832 at commit [`ad9b40a`](https://github.com/apache/spark/commit/ad9b40acf7daa28ba9b7faff279cfbad4b00c096).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

srowen commented:
@ScrapCodes any thoughts? otherwise LGTM

ScrapCodes commented:
I am inclined towards having sed replacement command which is compatible with POSIX (works both on bsd and GNU). I do not have a BSD system handy right now. :/  to help out.

srowen commented:
Yeah me too but I remember being unable to figure it out when I took a shot at it, and I assume you couldn't either. This at least uses GNU sed consistently.

ScrapCodes commented:
I am going to give it a shot, if you allow me some time.

mallman commented:
Sure thing. FYI, I'm leaving for Iceland tomorrow (Saturday), and I'll be away for two weeks. I will probably be incommunicado during this time. If you need something from me, can we put this PR on hold? Or if you find a cross-platform `sed` syntax that works, I won't mind if you just push that change. Let's keep the jackson and scaladoc fixes either way.

Cheers.

srowen commented:
@ScrapCodes go for it but do you have a better solution in sight? would it be reasonable to merge this as a step forward?

ScrapCodes commented:
@srowen I tried it, all we need to do is drop the `-i` flag, which in-place replaces the file. Instead we need to probably redirect the stream by correctly setting each filename using find or something. 

Unfortunately, I do not have a BSD system handy. I am going to try it out. Meanwhile if you could help, that would be great. Incase, nothing lands here in next 2 days feel free to merge this as interim solution.

sujkh85 commented:
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-8401] [Build] Scala version switching build enhancements (#6832)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

ScrapCodes commented:
BTW: Is there a way to report spam in github ?

srowen commented:
@ScrapCodes yes you can click the user link and report abuse. I have already done that but please go ahead too.

If you mean you need an OS X system, I think we can try it out. What change specifically needs to be made here?

ScrapCodes commented:
Try this as `dev/change-version-to-2.11.sh` on OSX with bsd sed.

```bash

# Note that this will not necessarily work as intended with non-GNU sed (e.g. OS X)
BASEDIR=$(dirname $0)/..

for f in `find $BASEDIR \( -name 'pom.xml' -type f -and -not -path '*target*' \) -print`; do
    echo -e "$f"
    sed -e 's/\(artifactId.*\)_2.10/\1_2.11/g' "$f" >"$f.tmp"
    cat "$f.tmp" >"$f"
    rm "$f.tmp"
done


# Also update <scala.binary.version> in parent POM
sed -e '0,/<scala\.binary\.version>2.10</s//<scala.binary.version>2.11</' "$BASEDIR/pom.xml" >"$BASEDIR/pom.xml.tmp"

cat "$BASEDIR/pom.xml.tmp" >"$BASEDIR/pom.xml"
rm "$BASEDIR/pom.xml.tmp"

```

ScrapCodes commented:
@srowen Did you get a chance to try the above out ? did it work ?

srowen commented:
@ScrapCodes that works as intended and doesn't leave the "-e" files around anymore, except that the update for `scala.binary.verison` does not work. This seemed to work:

```
sed -e 's/<scala\.binary\.version>2\.10</<scala.binary.version>2.11</' "$BASEDIR/pom.xml" >"$BASEDIR/pom.xml.tmp"
```

But I don't know if that's standard `sed` syntax now! seems simple though?

ScrapCodes commented:
Yes it looks simple, If you think this suggestion work for both gnu and bsd systems. Can we update the patch accordingly ?

srowen commented:
@mallman what do you think about these changes instead?

srowen commented:
@mallman are you able to look at this further or should maybe @ScrapCodes proceed with the changes he identified?

mallman commented:
@srowen I just returned from my vacation abroad and am catching up. Sorry for the wait. I'll take a look at this tomorrow. Cheers.

mallman commented:
I've spent some more time googling around this problem. Unsurprisingly, there's plenty of discussion/frustration around finding a cross-platform solution. There doesn't seem to be a magic bullet. In addition to the options we've discussed, I think there are two more worth considering. One is to use the `-i` flag but passing a backup file suffix like `sed -i.bak ...`. Then delete the `.bak` files. This solution is described here: http://stackoverflow.com/a/22084103. Another option is to use perl instead of sed.

That brings us to four approaches:

1. Look for GNU sed and exit if it can't be found. (what I did)
1. Write sed output to temporary files and move the temporary files to the original files. (what @ScrapCodes did)
1. Use the `-i.bak` syntax and delete the `.bak` files. This works on GNU and BSD sed, but maybe not XYZ or PDQ sed. The point being that `-i` in any form just isn't part of the posix standard.
1. Use `perl -i ...`, which is cross-platform.

At this point I would lean toward either (2) or (4), with (4) being the cleaner (easier) implementation. It also assumes the system has perl. Thoughts?

srowen commented:
I think I also like option 2. On OS X I actually am not sure whether I have GNU sed because of OS X or because of brew, so maybe slightly better not to have to rely on it, and to not have to detect it. I think this ends up simplifying the common script then?

ScrapCodes commented:
I was also inclined towards having an echo (Like the one in script above.). It gives the user necessary feedback, as to what changed.

mallman commented:
I'll work on a revision following along the lines of what @ScrapCodes did and push it to this PR. Incidentally, I was going to suggest we use `mktemp` to create the temp files, but then I realized that tool is different on bsd versus gnu systems.

mallman commented:
I've run into a roadblock. This syntax:

    sed -e '0,/<scala\.binary\.version>2.10</s//<scala.binary.version>2.11</' pom.xml

doesn't work with my Mac's sed. I'm not sure why. Maybe this is a bug in Yosemite's sed? I haven't tried other versions.

I spent a good deal of time this morning trying various alternatives without success. I need to put this aside for the day. If you guys find something that works please post.

Cheers.

srowen commented:
`sed -e 's/<scala\.binary\.version>2\.10</<scala.binary.version>2.11</' ...` worked for me on Yosemite. Is that not standard syntax too? that's what I always use with `sed`.

mallman commented:
The original code replaces only the first instance of `<scala.binary.version>2.10<` in the file, which is the desired behavior. The code you presented replaces all of them. You can more clearly see how that's a problem if you run that sed commend to change from version 2.10 to 2.11, then back again to 2.10, and then run `git diff pom.xml`.

srowen commented:
I thought that would be fine but, yeah it changes the 2.11 in the 2.11 profile to 2.10 on the way back, which is wrong. I fished around a little more and this seems to work as desired on OS X:

```
sed -e '1,/<scala\.binary\.version>2.10</s/<scala\.binary\.version>2.10</<scala.binary.version>2.11</' ...
```

That is, use 1 instead of 0, and repeat the pattern. See http://www.linuxtopia.org/online_books/linux_tool_guides/the_sed_faq/sedfaq4_004.html and http://stackoverflow.com/questions/148451/how-to-use-sed-to-replace-only-the-first-occurrence-in-a-file

Do you have GNU sed handy to confirm if that's portable? the page above suggests it is, since this won't occur on the first line.

mallman commented:
Thanks for the tip, @srowen. That works.

I now have a version following approach (2) which I've verified works on OS X with its built-in sed. I'll test on a GNU system later this afternoon or tomorrow. Assuming that works I'll push a new commit.

Commit:
Make the scala version change scripts cross-platform by restricting
ourselves to POSIX sed syntax instead of looking for GNU sed

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

mallman commented:
I've pushed a commit to implement the second strategy. I've tested this script successfully on OS X Yosemite and Ubuntu 14.

SparkQA commented:
[Test build #36943 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36943/consoleFull) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).

srowen commented on dev/change-version-to-2.11.sh:
```diff
@@ -17,10 +17,4 @@
-
-# Also update <scala.binary.version> in parent POM
-sed -i -e '0,/<scala\.binary\.version>2.10</s//<scala.binary.version>2.11</' $BASEDIR/pom.xml
+$(dirname $0)/change-scala-version.sh 2.10 2.11
```
@ScrapCodes should we just delete these two scripts and update any references to use the parameterized, new script? seems cleaner.

LGTM in any event.

SparkQA commented:
[Test build #36943 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36943/console) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1024 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1024/consoleFull) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).

SparkQA commented:
[Test build #1024 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1024/console) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1026 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1026/consoleFull) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).

SparkQA commented:
[Test build #1026 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1026/console) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1036 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1036/consoleFull) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).

SparkQA commented:
[Test build #1036 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1036/console) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).
 * This patch **fails MiMa tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1038 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1038/consoleFull) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).

SparkQA commented:
[Test build #1038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1038/console) for   PR 6832 at commit [`02296f2`](https://github.com/apache/spark/commit/02296f2c017b4bfb21ff286a0736a5df138eac2e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
Looking strong @mallman. @ScrapCodes how about we finish this by:

- Deleting the `change-version-to-*` scripts
- Updating `dev/create-release.sh` to reflect the new script
- Updating `docs/building-spark.md` too

mallman commented:
@srowen @ScrapCodes Let me know if you'd like me to take on those additional tasks. Cheers.

srowen commented:
@mallman I think you can go for it

srowen commented:
Ping @mallman I think you can make the additional changes. I don't mind just merging as is though, so will otherwise do that in a day or two.

mallman commented:
@srowen Sorry, I've been swamped. I think I can get this done by Saturday if you want to wait.

mallman commented:
@srowen I'm working on this now.

Commit:
Delete dev/change-version-to-*.sh, replacing them with single
dev/change-scala-version.sh script that takes a version as argument

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37848 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37848/consoleFull) for   PR 6832 at commit [`cde2f17`](https://github.com/apache/spark/commit/cde2f170d4562184eee7287c20b9115b3f457ebb).

mallman commented:
@srowen I've pushed a new commit to replace usage of `dev/change-version-to-*.sh` scripts with `dev/change-scala-version.sh <version>`. I also modified the latter so it takes a single argument instead of two. While this makes the current script more tied to version 2.10 and 2.11 specifically, I believe the simplicity in usage obtained is worth deferring the extra complexity of programmatically determining the "from version" when we begin support for 2.12.

SparkQA commented:
[Test build #37848 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37848/console) for   PR 6832 at commit [`cde2f17`](https://github.com/apache/spark/commit/cde2f170d4562184eee7287c20b9115b3f457ebb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

srowen commented:
Looks good to me. @ScrapCodes any more thoughts?

ScrapCodes commented:
Yeah, Looks good !

srowen commented:
Merged into master
